File: src/D3D11/D3D11Instance.cpp
#include "D3D11Instance.h"
#include "D3D11Utils.h"

namespace grabs::D3D11
{
    D3D11Instance::D3D11Instance(const InstanceInfo& info)
    {
        D3D11_CHECK_RESULT(CreateDXGIFactory1(__uuidof(IDXGIFactory1), reinterpret_cast<void**>(&Factory)));
    }

    D3D11Instance::~D3D11Instance()
    {
        Factory->Release();
    }

    std::unique_ptr<Device> D3D11Instance::CreateDevice(Surface* surface, uint32_t adapterIndex)
    {
        return nullptr;
    }

    std::vector<Adapter> D3D11Instance::EnumerateAdapters()
    {
        std::vector<Adapter> adapters;

        IDXGIAdapter1* dxgiAdapter;
        for (UINT i = 0; Factory->EnumAdapters1(i, &dxgiAdapter) != DXGI_ERROR_NOT_FOUND; i++)
        {
            DXGI_ADAPTER_DESC1 desc;
            D3D11_CHECK_RESULT(dxgiAdapter->GetDesc1(&desc));

            std::wstring wName = desc.Description;
            const std::string name(wName.begin(), wName.end());

            Adapter adapter
            {
                .Index = i,
                .Name = name,
                .Memory = desc.DedicatedVideoMemory,
                .Supports = AdapterSupports
                {
                    .GeometryShader = true,
                    .Anisotropy = true,
                    .MaxAnisotropy = D3D11_MAX_MAXANISOTROPY
                }
            };

            adapters.push_back(adapter);
        }

        return adapters;
    }
}

File: src/grabs.Graphics/D3D11/D3D11Instance.cs
using grabs.Core;
using Vortice.DXGI;

namespace grabs.Graphics.D3D11;

internal sealed class D3D11Instance : Instance
{
    private readonly bool _debug;
    
    public readonly IDXGIFactory1 Factory;
    
    public override Backend Backend => Backend.D3D11;

    public D3D11Instance(ref readonly InstanceInfo info)
    {
        _debug = info.Debug;
        
        GrabsLog.Log("Creating DXGI factory.");
        Factory = DXGI.CreateDXGIFactory1<IDXGIFactory1>();
    }
    
    public override Adapter[] EnumerateAdapters()
    {
        List<Adapter> adapters = new List<Adapter>();
        
        for (uint i = 0; Factory.EnumAdapters1(i, out IDXGIAdapter1 adapter).Success; i++)
        {
            AdapterDescription1 desc = adapter.Description1;
            
            string name = desc.Description;

            AdapterType type = AdapterType.Dedicated;
            if (desc.Flags.HasFlag(AdapterFlags.Software))
                type = AdapterType.Software;

            ulong memory = desc.DedicatedVideoMemory;
            
            adapters.Add(new Adapter(adapter.NativePointer, i, name, type, memory));
        }

        return adapters.ToArray();
    }
    
    public override Device CreateDevice(Surface surface, Adapter? adapter = null)
    {
        IDXGIAdapter1 dxgiAdapter;
        
        if (adapter is { } givenAdapter)
            dxgiAdapter = new IDXGIAdapter1(givenAdapter.Handle);
        else
        {
            Adapter[] adapters = EnumerateAdapters();
            dxgiAdapter = new IDXGIAdapter1(adapters[0].Handle);
        }

        return new D3D11Device(Factory, dxgiAdapter, _debug);
    }
    
    public override Surface CreateSurface(in SurfaceInfo info)
    {
        return new D3D11Surface(in info);
    }
    
    public override void Dispose()
    {
        Factory.Dispose();
    }
}